src/lazy/encoder/binary/v1_1/flex_uint.rs

```diff
@@ -0,0 +1,229 @@
+    while leading_zeros < BITS_PER_U64 {
+        let magnitude_bits_needed = 64 - leading_zeros;
+        cache[leading_zeros] =
+            ((magnitude_bits_needed + BITS_PER_ENCODED_BYTE - 1) / BITS_PER_ENCODED_BYTE) as u8;
+        leading_zeros += 1;
+    }
+    // Special case; 64 leading zeros means the value is zero. We need a byte to represent it.
+    cache[64] = 1;
+    cache
+}
+
+static BYTES_NEEDED_CACHE: [u8; 65] = init_bytes_needed_cache();
+
+/// An Ion 1.1 encoding primitive that represents a variable-length unsigned integer.
+#[derive(Debug)]
+pub struct FlexUInt {
+    value: u64,
+    size_in_bytes: usize,
+}
+
+impl FlexUInt {
+    pub fn new(size_in_bytes: usize, value: u64) -> Self {
+        Self {
+            value,
+            size_in_bytes,
+        }
+    }
+
+    #[inline]
+    pub fn write_u64<W: Write>(output: &mut W, value: u64) -> IonResult<usize> {
+        let leading_zeros = value.leading_zeros();
+        let num_encoded_bytes = BYTES_NEEDED_CACHE[leading_zeros as usize] as usize;
+        if num_encoded_bytes <= 8 {
+            let flag_bits = 1u64 << (num_encoded_bytes - 1);
+            // Left shift the value to accommodate the trailing flag bits and then OR them together
+            let encoded_value = (value << num_encoded_bytes) | flag_bits;
+            output.write_all(&encoded_value.to_le_bytes()[..num_encoded_bytes])?;
+            return Ok(num_encoded_bytes);
+        }
+        Self::write_large_u64(output, value, num_encoded_bytes)
+    }
+
+    /// Helper method that encodes a signed values that require 9 or 10 bytes to represent.
+    /// This code path is rarely used and requires more instructions than the common case.
+    /// Keeping it in a separate method allows the common case to be inlined in more places.
+    fn write_large_u64<W: Write>(
+        output: &mut W,
+        value: u64,
+        encoded_size_in_bytes: usize,
+    ) -> IonResult<usize> {
+        match encoded_size_in_bytes {
+            9 => {
+                // When combined with the continuation flags, the value is too large to be encoded in
+                // a u64. It will be nine bytes in all.
+                //
+                // Set up a stack-allocated buffer to hold our encoding. This allows us to call
+                // `output.write_all()` a single time.
+                let mut buffer: [u8; 9] = [0; 9];
+
+                // The first byte will always be 0x00, indicating that 8 more bytes follow.
+                //
+                // We need to leave a `1` in the low bit of the second byte to be the End flag. Because
+                // we need fewer than 64 bits for magnitude, we can encode the remainder of the data
+                // in a u64.
+                let encoded_value = (value << 1) + 1; // Leave a trailing `1` in the lowest bit
+                buffer[1..].copy_from_slice(&encoded_value.to_le_bytes()[..]);
+                output.write_all(buffer.as_slice())?;
```

Comment:
Does it really make a big difference being able to write all 9 bytes at one time? It seems to me that it would be simpler to implement like this:

```rust
const zero: [u8; 1] = [0];
output.write_all(&zero)?;
let encoded_value = (value << 1) + 1;
output.write_all(&encoded_value.to_le_bytes()[..])?;
```

Reply:
I adapted this impl from a prototype implementation I had lying around. In benchmarking the prototype, I found that avoiding two calls to `write_all` was a minor net benefit because it avoided the additional error handling branches involved. I agree that the version you wrote is simpler, and this code path may not benchmark the same way as my prototype.

I'll try the simpler impl and see if it moves the needle either way in the benchmarks.

Reply:
Actually, as I've looked at it more, I've realized that the cost of copying 8 bytes is probably less than the cost of a function call, and there's not much difference in terms of readability. Don't spend more than a few minutes trying it out if you try it at all.